File: knowledge_mcp/src/knowledge_mcp/models/tag.py
"""TagNode データモデル（実装指示書 5.6 / ADR-0006、IMPL-202608060837 4.3 で description/aliases 拡張）。"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class TagNode:
    id: int
    name: str
    parent_tag_id: Optional[int] = None
    description: Optional[str] = None
    # タグフォルダ（分類表示専用メタデータ、ADR-0072）。0 または 1 件（None=未分類）。
    # parent_tag_id による is-a 階層とは独立し、search_knowledge の検索には一切使わない。
    folder_id: Optional[int] = None
    # 兄弟集合（同じ parent_tag_id を持つノード）内でのローカルな表示順序（ADR-0074）。
    # 値が小さいほど先に表示される。表示専用属性であり検索・スコアには関与しない。
    display_order: Optional[float] = None
    # alias 一覧。[{"id": int, "alias": str}, ...]
    aliases: List[dict] = field(default_factory=list)
    children: List["TagNode"] = field(default_factory=list)

    def to_dict(self, include_children: bool = True) -> dict:
        """MCP出力用の辞書へ変換する。

        include_children=False の場合は単一タグ（create_tag / rename_tag / move_tag 等の戻り値）
        として children を含めない。
        """
        d = {
            "id": self.id,
            "name": self.name,
            "parent_tag_id": self.parent_tag_id,
            "description": self.description,
            "folder_id": self.folder_id,
            "display_order": self.display_order,
            "aliases": [dict(a) for a in self.aliases],
        }
        if include_children:
            d["children"] = [c.to_dict() for c in self.children]
        return d

File: front_dev/src/domain/admin/tag.ts
// タグ管理の型定義（IMPL-202608060837 T16 / Knowledge MCP の TagNode に対応）

export interface TagAlias {
  id: number;
  alias: string;
}

export interface TagNode {
  id: number;
  name: string;
  parent_tag_id: number | null;
  description: string | null;
  // タグフォルダ（分類表示専用メタデータ, ADR-0072）。null は未分類。
  folder_id: number | null;
  // 兄弟集合内の表示順序（ADR-0074）。並べ替えの境界判定は配列位置で行うため
  // 値そのものは通常使わないが、デバッグ・将来のD&D拡張のため保持する。
  display_order: number;
  aliases: TagAlias[];
  children: TagNode[];
}

// タグ・タグフォルダの並べ替え（ADR-0074）。"up"=1つ上へ、"down"=1つ下へ。
export interface TagReorderRequest {
  direction: "up" | "down";
}

export interface TagCreateRequest {
  name: string;
  parent_tag_id?: number | null;
  description?: string | null;
  folder_id?: number | null;
}

// name / description / parent_tag_id / folder_id のいずれか複数を指定できる（送った項目のみ更新）。
// folder_id は送った場合のみ更新（null 指定でタグフォルダを解除）。
export interface TagUpdateRequest {
  name?: string;
  description?: string | null;
  parent_tag_id?: number | null;
  folder_id?: number | null;
}

// タグフォルダマスタ（分類表示専用メタデータ, ADR-0072 で新設、ADR-0073 で階層化）。
// list_tag_folders は parent_folder_id による木構造（children にネスト）で返す。
// name は表示用ラベルで一意性を持たない（同名フォルダが複数存在しうる, ADR-0073）。
export interface TagFolder {
  id: number;
  name: string;
  description: string | null;
  parent_folder_id: number | null;
  // 兄弟集合内の表示順序（ADR-0074）。
  display_order: number;
  children: TagFolder[];
}

export interface TagFolderCreateRequest {
  name: string;
  description?: string | null;
  // 親フォルダ（ADR-0073）。省略/null でルートフォルダとして作成する。
  parent_folder_id?: number | null;
}

export interface TagFolderUpdateRequest {
  name: string;
  description?: string | null;
}

// タグフォルダの移動（ADR-0073）。null でルートフォルダへ移動する。
export interface TagFolderMoveRequest {
  new_parent_folder_id: number | null;
}

// タグ一括インポートの結果（IMPL-202608261630 T4 / ADR-0061）
export interface TagImportRowResult {
  row: number;
  status: "success" | "error";
  tag_id?: number;
  error?: string;
}

export interface TagImportResponse {
  total: number;
  success_count: number;
  results: TagImportRowResult[];
}
